### PR TITLE
[FEATURE] Add new seminar single view hook

### PR DESCRIPTION
**Pull request by mk-mxp**
As requested by @oliverklee in #241

There is these open questions:

- Is the interface feature complete? Or add methods for next day events list / other dates list, too?
- Should I deprecate the old hook? If so, like for new RegistrationEmailHookInterface?

As a side note: One of the CsvDownloaderTests failed when I had an error (threw an unexpected Exception in GeneralUtility) in the test I added:
```
There was 1 failure:

1) Tx_Seminars_Tests_Unit_Csv_CsvDownloaderTest::createAndOutputListOfRegistrationsForInexistentEventUidReturn404
Failed asserting that 'Content-disposition: attachment; filename=' contains "404".

/home/michael/MXP/ext-seminars/Tests/LegacyUnit/Csv/CsvDownloaderTest.php:823
...
```

Should I open an issue for that? Or will the test be re-implemented anyways soon?

### CODE REVIEW

**Comment by oliverklee** [comment]
> Should I deprecate the old hook? If so, like for new RegistrationEmailHookInterface?

Yes, please.

**Comment by oliverklee** [comment]
About the failed test:

Unfortunately, the legacy tests are terrible brittle as they depend on the data in your current TYPO3 installation. If the problem is not a hiccup, but constantly occurs when you run the corresponding test case 3 times, then please open an issue.

File: Classes/Hooks/Interfaces/SeminarSingleView.php [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+declare(strict_types=1);
+
+namespace OliverKlee\Seminars\Hooks\Interfaces;
+
+use OliverKlee\Seminars\Hooks\Interfaces\Hook;
+
+/**
+ * Use this interface for hooks concerning the seminar single view.
+ *
+ * It superseeds the deprecated EventSingleView interface.
```
Comment (oliverklee) [inline Classes/Hooks/Interfaces/SeminarSingleView.php]: typo: superseeds -> supersedes

**Comment by oliverklee** [comment]
> Is the interface feature complete? Or add methods for next day events list / other dates list, too?

I'd say we only implement what you need at this time (YAGNI).

For the interface, I'd prefer for the hook to received an `TemplateHelper` instance (in this case, the  `DefaultController` instance) and the event instance so that the hook can work on the template with the data of the event.

**Comment by mk-mxp** [comment]
**YAGNI**

This is one of the instances where YAGNI is the wrong principle. The hook, as a design principle,  **shall** enable usages **not needed by me**. Usages, not even thought of by me.

IMO the question is: Is it enough to have next day events list / other dates list hooks via the seminar list hook, or should it be included in this interface, too (so my implementing class can simply use protected vars to pass info to the list hooks instead of having static vars or singletons).

Anyways, besides that, I don't need it. But it is impossible to extend the interface later without being incompatible immediately.

**Comment by mk-mxp** [comment]
> For the interface, I'd prefer for the hook to received an `TemplateHelper` instance (in this case, the `DefaultController` instance) and the event instance so that the hook can work on the template with the data of the event.

Again a case of unnecessarily narrow restrictions. Why cut the hook user off from `getConfigurationService()`, `isLoggedIn()`, `getLoginLink()` or any other public method of `DefaultController`? This **is** the main reason why I re-implemented your hooks in the first place. `$controller->isLoggedIn()` saved my life.

**Comment by oliverklee** [comment]
Generally, concerning hooks: Usually, I (or users) implement hooks when someone needs a particular hook and requests it.

Concerning the next day/other dates list: I currently see no benefit of having a dedicated hook over the general list view hook.

Concerning passing the `DefaultController` instance to hooks: This class is one big mess of code: It combines almost all views and has no API that is intended to be used from the outside. Unless XClasses, I expect hook providers to have a stable API with proper deprecations. How can we handle this?

**Comment by mk-mxp** [comment]
> Concerning passing the `DefaultController` instance to hooks: This class is one big mess of code: It combines almost all views and has no API that is intended to be used from the outside. Unless XClasses, I expect hook providers to have a stable API with proper deprecations. How can we handle this?

From the hook user perspective: There is no "stability guaranteed" API to any of those classes (yet). Meaning: Anounce major changes for 3.0/4.0/5.0 in the hook's interface (the most likely place for a hook user to look for a change). Breaking changes are never welcome. When you are confident with architecture again, stability will be there.

With any major version I may or may not have to re-implement my hooked-in classes. Just the same as I have to do for Typo3 LTS. Knowing that seminars will have Fluid templates one day makes `TemplateHelper` deprecated, too. And whatever is passed to the methods: When the interface is to be changed (e.g. for a namespace), the whole hook becomes incompatible. A streamlined implementation of a controller will, for sure, not be named `\Tx_Seminars_Frontend_DefaultController` anymore.

In short: Don't care too much about staying compatible when you need to cleanup a mess. Splitting controllers will break a lot more. Fluid templates will, DB-Handling will, too.

File: Classes/FrontEnd/DefaultController.php [hunk 2]
```diff
@@ -405,13 +416,17 @@ protected function getListViewHooks(): array
...
+     * @deprecated will be removed in seminars 4; use ->getSingleViewHookProvider() instead
+
      */
     protected function getSingleViewHooks(): array
     {
         if (!$this->singleViewHooksHaveBeenRetrieved) {
             $hookClasses = $GLOBALS['TYPO3_CONF_VARS']['EXTCONF']['seminars']['singleView'];
             if (is_array($hookClasses)) {
                 foreach ($hookClasses as $hookClass) {
+                    GeneralUtility::deprecationLog($hookClass . ' - since seminars 3.0, interface \\Tx_Seminars_Interface_Hook_EventSingleView will be removed in seminars 4.0');
```
Comment (oliverklee) [inline Classes/FrontEnd/DefaultController.php]: This line is longer than 120 characters (the hard limit of PSR-12, although I have not enabled that sniff yet). Could you please split it up?

File: Classes/FrontEnd/DefaultController.php [hunk 3]
```diff
@@ -430,6 +445,20 @@ protected function getSingleViewHooks(): array
...
     }
 
+    /**
+     * Gets the hook provider for the single view.
+     *
+     * @return HookProvider
+     */
+    protected function getSingleViewHookProvider(): HookProvider
+    {
+        if (!$this->singleViewHookProvider instanceof HookProvider) {
```
Comment (oliverklee) [inline Classes/FrontEnd/DefaultController.php]: Please use `=== null` instead to make it clear that this is about a null check, not about avoiding instances of other classes.

File: Classes/Hooks/Interfaces/SeminarSingleView.php [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+declare(strict_types=1);
+
+namespace OliverKlee\Seminars\Hooks\Interfaces;
+
+use OliverKlee\Seminars\Hooks\Interfaces\Hook;
+
+/**
+ * Use this interface for hooks concerning the seminar single view.
+ *
+ * It supersedes the deprecated EventSingleView interface.
```
Comment (oliverklee) [inline Classes/Hooks/Interfaces/SeminarSingleView.php]: Please use backticks around the class name.

File: Classes/Hooks/Interfaces/SeminarSingleView.php [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+ *
+ * @author Michael Kramer <m.kramer@mxp.de>
+ */
+interface SeminarSingleView extends Hook
+{
+    /**
+     * Modifies the seminar details view.
+     *
+     * This function will be called for all types of seminars.
```
Comment (oliverklee) [inline Classes/Hooks/Interfaces/SeminarSingleView.php]: This is a bit ambiguous. I propose adding "(single events, topics, and dates)".

File: Tests/LegacyUnit/FrontEnd/DefaultControllerTest.php [hunk 6]
```diff
@@ -1260,6 +1261,26 @@ public function singleViewCallsModifyEventSingleViewHook()
...
+    public function singleViewCallsHookSeminarSingleViewModifySingleView()
+    {
+        $this->subject->setConfigurationValue('what_to_display', 'single_view');
+
+        /** @var \Tx_Seminars_Model_Event $event */
+        $event = \Tx_Oelib_MapperRegistry::get(\Tx_Seminars_Mapper_Event::class)->find($this->seminarUid);
+        $hookedInObject = $this->createMock(SeminarSingleView::class);
+        $hookedInObject->expects(self::once())->method('modifySingleView')->with($this->subject);
+
+        $hookedInClass = get_class($hookedInObject);
```
Comment (oliverklee) [inline Tests/LegacyUnit/FrontEnd/DefaultControllerTest.php]: Please add a backslash before `get_class` to improve performance (once this test gets namespaced).

File: Tests/LegacyUnit/FrontEnd/DefaultControllerTest.php [hunk 7]
```diff
@@ -1260,6 +1261,26 @@ public function singleViewCallsModifyEventSingleViewHook()
...
+        $GLOBALS['TYPO3_CONF_VARS']['EXTCONF']['seminars'][SeminarSingleView::class][] = $hookedInClass;
+        GeneralUtility::addInstance($hookedInClass, $hookedInObject);
+
+        $this->subject->piVars['showUid'] = $this->seminarUid;
```
Comment (oliverklee) [inline Tests/LegacyUnit/FrontEnd/DefaultControllerTest.php]: Please cast the UID to a string here as the pivars are strings by default.

File: Classes/FrontEnd/DefaultController.php [hunk 8]
```diff
@@ -405,13 +416,17 @@ protected function getListViewHooks(): array
      * @throws \UnexpectedValueException
      *          if there are registered hook classes that do not implement the
      *          \Tx_Seminars_Interface_Hook_EventSingleView interface
+     *
+     * @deprecated will be removed in seminars 4; use ->getSingleViewHookProvider() instead
+
```
Comment (oliverklee) [inline Classes/FrontEnd/DefaultController.php]: This blank line seems to have snuck in.

File: Classes/FrontEnd/DefaultController.php (hunk 3 above)
Comment (mk-mxp) [reply Classes/FrontEnd/DefaultController.php]: I'll do that, but: https://rskuipers.com/entry/from-assumptions-to-assertions

**Comment by mk-mxp** [comment]
> And could you also please add the changelog lines (one for the new hook, and one for the deprecation (both with the same PR ID)?

Will add them as last, avoids conflicts when rebasing...

**Comment by oliverklee** [comment]
@mk-mxp Thanks for the article recommendation about null checks. I've read the article, learned something from it and will heed its advice in the future.

**Comment by oliverklee** [comment]
@mk-mxp I forgot one thing here: Could you please document the new hook in the extension manual? Thanks!
